Publications
============

Citation
--------

When citing PyHMMER, please additionally cite HMMER, which developed the actual
method and is wrapped with PyHMMER. You can for instance cite both papers as:
*PyHMMER (Larralde et al., 2022), a Python library binding to HMMER (Eddy, 2011).*

Sean Eddy states in the `HMMER manual <http://eddylab.org/software/hmmer/Userguide.pdf>`_
that you should cite the HMMER website if possible, and only resort to the 2011
publication (:doi:`10.1371/journal.pcbi.1002195`) for journals that cause
problems. You can grab a `BibTex file <_static/bibtex/citation.bib>`_ with
the different publications.


Users
-----

PyHMMER has been used in the following research works:

- Carroll, L.M., Larralde, M., Fleck, J. S., Ponnudurai, R., Milanese, A., Cappio, E., Zeller, G., 2021. *Accurate de novo identification of biosynthetic gene clusters with GECCO* (preprint). bioRxiv. :doi:`10.1101/2021.05.03.442509`.
- Lienhard, M., van den Beucken, T., Timmermann, B., Hochradel, M., Boerno, S., Caiment, F., Vingron, M., Herwig, R., 2021. *Long-read transcriptome sequencing analysis with IsoTools* (preprint). Genetics. :doi:`10.1101/2021.07.13.452091`.
- Kautsar, S.A., van der Hooft, J.J.J., de Ridder, D., Medema, M.H., 2021. *BiG-SLiCE: A highly scalable tool maps the diversity of 1.2 million biosynthetic gene clusters*. GigaScience 10, giaa154. :doi:`10.1093/gigascience/giaa154`.
- Konstantinidis, K., Ruiz-Perez, C., Gerhardt, K., Rodriguez-R, L., Jain, C., Tiedje, J., Cole, J., 2022. *FastAAI: Efficient Estimation of Genome Average Amino Acid Identity and Phylum-level relationships using Tetramers of Universal Protein* (preprint). ResearchSquare. :doi:`10.21203/rs.3.rs-1459378/v1`.
